File: src/main/resources/sse.js
const mode = '$mode$';

document.addEventListener("DOMContentLoaded", () => {
    const bookmarkList = document.getElementById('movie-list');

    const eventSource = new EventSource('/sse-stream?mode=' + mode);

    eventSource.onmessage = function(event) {
        if (event.data === "heartbeat") return;
        const data = JSON.parse(event.data);
        console.log('Received SSE:', data);
        const elements = document.querySelectorAll(`#${data.kind.toLowerCase()}-${data.id}`);
        const className = data.kind.toLowerCase() === 'bookmark' ? 'bookmarked' : 'watched';
        switch (data.type) {
            case 'de.amklee.monomovie.components.SseEvent.Add':
                if (elements.length > 0) {
                    elements.forEach(element => element.classList.add(className));
                } else if (bookmarkList && data.insert) {
                    bookmarkList.prepend(document.parseHtml(data.body));
                }
                break;
            case 'de.amklee.monomovie.components.SseEvent.Remove':
                elements.forEach((element) => element.classList.remove(className));
                break;
            default:
                console.warn('Unknown event type:', data.type);
        }
    };

    eventSource.onerror = function(err) {
        //TODO reconnect on network change (or similar)
        console.error('EventSource failed:', err);
        eventSource.close();
    };
});

File: src/main/kotlin/de/amklee/monomovie/components/Sse.kt
package de.amklee.monomovie.components

import de.amklee.monomovie.service.CachedMovies
import de.amklee.monomovie.service.Event
import de.amklee.monomovie.util.buildULHtml
import kotlinx.serialization.Serializable

@Serializable
sealed interface SseEvent {
    val kind: Kind
    val id: String

    @Serializable data class Add(
        override val kind: Kind,
        override val id: String,
        val body: String,
        val insert: Boolean
    ) : SseEvent

    @Serializable data class Remove(
        override val kind: Kind,
        override val id: String
    ) : SseEvent
}

enum class Kind {
    BOOKMARK, WATCHED
}

enum class Mode {
    SEARCH, WATCHED, OVERVIEW, ROULETTE, OFFERS
}

suspend fun convertBookmarkSse(event: Event, mode: Mode, kind: Kind): SseEvent? {
    if (mode == Mode.ROULETTE && kind == Kind.BOOKMARK) return null
    val movie = CachedMovies.get(event.id) ?: return null
    return when (event) {
        is Event.Added -> SseEvent.Add(
            id = event.id,
            kind = kind,
            body = buildULHtml {
                if (mode == Mode.OVERVIEW) SelectableMovieListItem(movie)
                else MovieListItem(movie)
            },
            insert = when (mode) {
                Mode.SEARCH -> false
                Mode.WATCHED -> kind == Kind.WATCHED
                Mode.OVERVIEW -> kind == Kind.BOOKMARK
                Mode.ROULETTE -> false
                Mode.OFFERS -> false
            }
        )
        is Event.Removed -> {
            SseEvent.Remove(
                id = event.id,
                kind = kind
            )
        }
    }
}
